roxie/ccd/ccdquery.cpp

```diff
@@ -1696,33 +1789,35 @@ static void checkWorkunitVersionConsistency(const IQueryDll *dll)
 
 extern IQueryFactory *createServerQueryFactory(const char *id, const IQueryDll *dll, const IRoxiePackage &package, const IPropertyTree *stateInfo, bool isDynamic, bool forceRetry)
 {
-    CriticalBlock b(CQueryFactory::queryCreateLock);
     IArrayOf<IResolvedFile> queryFiles; // Note - these should stay in scope long enough to ensure still cached when (if) query is loaded for real
     hash64_t hashValue = CQueryFactory::getQueryHash(id, dll, package, stateInfo, queryFiles, isDynamic);
-    IQueryFactory *cached = getQueryFactory(hashValue, 0);
-    if (cached && !(cached->loadFailed() && (reloadRetriesFailed || forceRetry)))
-    {
-        ::Release(dll);
-        return cached;
-    }
-    if (dll && !selfTestMode)
-    {
-        checkWorkunitVersionConsistency(dll);
-        Owned<ISharedOnceContext> sharedOnceContext;
-        IPropertyTree *workflow = dll->queryWorkUnit()->queryWorkflowTree();
-        if (workflow && workflow->hasProp("Item[@mode='once']"))
-            sharedOnceContext.setown(new CSharedOnceContext);
-        Owned<CRoxieServerQueryFactory> newFactory = new CRoxieServerQueryFactory(id, dll, dynamic_cast<const IRoxiePackage&>(package), hashValue, sharedOnceContext, isDynamic);
-        newFactory->load(stateInfo);
-        if (sharedOnceContext && preloadOnceData)
+    Owned<CQueryFactory> ret;
+    {
+        CriticalBlock b(CQueryFactory::queryCacheCrit);
+        ret.setown(CQueryFactory::getCachedQuery(hashValue, 0));
+        if (ret && ret->loadFailed() && (reloadRetriesFailed || forceRetry))  // MORE - is there a race on loadFailed?
         {
-            Owned<StringContextLogger> logctx = new StringContextLogger(id); // NB may get linked by the onceContext
-            sharedOnceContext->checkOnceDone(newFactory, *logctx);
+            ret.clear();
+        }
+		if (ret)
+		    ::Release(dll);
+        else
+        {
+            if (dll && !selfTestMode)
+            {
+                checkWorkunitVersionConsistency(dll);
+                Owned<ISharedOnceContext> sharedOnceContext;
+                IPropertyTree *workflow = dll->queryWorkUnit()->queryWorkflowTree();
+                if (workflow && workflow->hasProp("Item[@mode='once']"))
+                    sharedOnceContext.setown(new CSharedOnceContext);
+                ret.setown(new CRoxieServerQueryFactory(id, dll, dynamic_cast<const IRoxiePackage&>(package), hashValue, sharedOnceContext, isDynamic));
```

Comment:
Not directly connected to this PR, but this dynamic_cast is unnecessary (also line 1816).  If it isn't optimized away it is a waste...